discovery/aws/ec2_test.go

```diff
@@ -0,0 +1,472 @@
+
+	return expected
+}
+
+func generateRefreshIpv4(d *EC2Discovery) []*targetgroup.Group {
+	ec2Data := &d.ec2.(*mockEC2Client).ec2Data
+
+	azIdx := generateTestInstanceData(ec2Data)
+
+	ec2Data.instances[0].SetSubnetId(ec2Data.azIDs[azIdx])
+	ec2Data.instances[0].SetVpcId(fmt.Sprintf("vpc-%s", randString(8)))
+
+	enis := make([]*ec2.InstanceNetworkInterface, 4)
+
+	// interface in the primary subnet
+	enis[0] = &ec2.InstanceNetworkInterface{}
+	enis[0].SetIpv6Addresses([]*ec2.InstanceIpv6Address{})
+	enis[0].SetSubnetId(*ec2Data.instances[0].SubnetId)
+
+	// interface without any subnet
+	enis[1] = &ec2.InstanceNetworkInterface{}
+	enis[1].SetIpv6Addresses([]*ec2.InstanceIpv6Address{})
+	// azIdx + 1 modulo #azs
+	secondAzID := (azIdx + 1) % len(ec2Data.azIDs)
+	enis[1].SetSubnetId(ec2Data.azIDs[secondAzID])
+
+	// interface in another subnet
+	enis[2] = &ec2.InstanceNetworkInterface{}
+	enis[2].SetIpv6Addresses([]*ec2.InstanceIpv6Address{})
+
+	// inteface in the primary subnet
+	enis[3] = &ec2.InstanceNetworkInterface{}
+	enis[3].SetIpv6Addresses([]*ec2.InstanceIpv6Address{})
+	enis[3].SetSubnetId(*ec2Data.instances[0].SubnetId)
+
+	ec2Data.instances[0].SetNetworkInterfaces(enis)
+
+	labels := generateExpectedLabels(ec2Data, azIdx, d.cfg.Port)
+
+	labels["__meta_ec2_primary_subnet_id"] = model.LabelValue(*ec2Data.instances[0].SubnetId)
+	labels["__meta_ec2_subnet_id"] = model.LabelValue(fmt.Sprintf(",%s,%s,", ec2Data.azIDs[azIdx], ec2Data.azIDs[secondAzID]))
+	labels["__meta_ec2_vpc_id"] = model.LabelValue(*ec2Data.instances[0].VpcId)
+
+	expected := make([]*targetgroup.Group, 1)
+	expected[0] = &targetgroup.Group{
+		Source: ec2Data.region,
+	}
+	expected[0].Targets = append(expected[0].Targets, labels)
+
+	return expected
+}
+
+func generateRefreshIpv6(d *EC2Discovery) []*targetgroup.Group {
+	ec2Data := &d.ec2.(*mockEC2Client).ec2Data
+
+	azIdx := generateTestInstanceData(ec2Data)
+
+	ec2Data.instances[0].SetSubnetId(ec2Data.azIDs[azIdx])
+	ec2Data.instances[0].SetVpcId(fmt.Sprintf("vpc-%s", randString(8)))
+
+	attachments := make([]*ec2.InstanceNetworkInterfaceAttachment, 4)
+	enis := make([]*ec2.InstanceNetworkInterface, 4)
+	ips := make([][]*ec2.InstanceIpv6Address, 4)
+
+	// interface without primary IPv6
+	attachments[0] = &ec2.InstanceNetworkInterfaceAttachment{}
+	attachments[0].SetDeviceIndex(3)
+
+	ips[0] = make([]*ec2.InstanceIpv6Address, 1)
+	ips[0][0] = &ec2.InstanceIpv6Address{}
+	ips[0][0].SetIpv6Address(randIpv6())
+	ips[0][0].SetIsPrimaryIpv6(false)
+
+	enis[0] = &ec2.InstanceNetworkInterface{}
+	enis[0].SetAttachment(attachments[0])
+	enis[0].SetIpv6Addresses(ips[0])
+	enis[0].SetSubnetId(*ec2Data.instances[0].SubnetId)
+
+	// interface with primary IPv6
+	attachments[1] = &ec2.InstanceNetworkInterfaceAttachment{}
+	attachments[1].SetDeviceIndex(2)
+
+	ips[1] = make([]*ec2.InstanceIpv6Address, 2)
+	ips[1][0] = &ec2.InstanceIpv6Address{}
+	ips[1][0].SetIpv6Address(randIpv6())
```

Comment:
I think that instead of using setters/field assignments, composite literal initialization would be more readable, WDYT?

Reply:
AWS SDK expects pointers, even on boolean or int64 values. IDK why.
As you requested I modified the tests to use only static data but I had to add ugly addrX functions to get pointers (that's how Go calls them) on literals.
So this code is gone by now.